util/generate_compilation_db.py

```diff
@@ -194,12 +194,60 @@ def build_compile_commands(
 
 
 def main(args):
+    # These C compiler flags are required for on-device targets.
+    REQUIRED_DEVICE_CC_FLAGS = set([
+        "-Wall",
+        "-Werror",
+        "-Wextra",
+        "-Wgnu",
+        "-Wimplicit-fallthrough",
+        "-Winvalid-pch",
+        "-Wno-covered-switch-default",
+        "-Wno-error=unused-function",
+        "-Wno-gnu-zero-variadic-macro-arguments",
+        "-Wno-missing-field-initializers",
+        "-Wno-sign-compare",
+        "-Wno-unused-parameter",
+        "-Wstrict-prototypes",
+        "-Wswitch-default",
+        "-Wconversion",
+    ])
+
     paths = PathBuilder(os.path.realpath(__file__))
 
-    device_commands, device_unittest_commands = build_compile_commands(
-        paths, device_build=True)
-    host_commands, host_unittest_commands = build_compile_commands(
-        paths, device_build=False)
+    device_commands = []
+    device_unittest_commands = []
+    host_commands = []
+    host_unittest_commands = []
+
+    if not args.audit_device_cc_flags:
+        device_commands, device_unittest_commands = build_compile_commands(
+            paths, device_build=True)
+        host_commands, host_unittest_commands = build_compile_commands(
+            paths, device_build=False)
+    else:
+        all_device_commands, _ = build_compile_commands(paths,
+                                                        device_build=True)
+
+        troubled_sources = []
+
+        for command in all_device_commands:
+            arguments = command["arguments"]
+            source_file = command["file"]
+            assert "-march=rv32imc" in arguments
+
+            missing_flags = REQUIRED_DEVICE_CC_FLAGS - set(arguments)
+            if not missing_flags:
+                continue
+
+            troubled_sources.append(source_file)
+
+            command["missing_cc_flags"] = sorted(list(missing_flags))
```

Comment:
thanks for `sorted(list(...))`

Reply:
Now that you mentioned it, I realize I should update the gist linked at the top of the PR description.

Reply:
Done.